lgc/patch/PatchBoundsCheckMemory.cpp

```diff
@@ -0,0 +1,195 @@
+// @param getElemPtr : "GetElementPtr" instruction
+void PatchBoundsCheckMemory::visitGetElementPtrInst(GetElementPtrInst &getElemPtr) {
+  auto pointerTy = getElemPtr.getPointerOperand()->getType();
+  if (pointerTy->getPointerAddressSpace() != ADDR_SPACE_PRIVATE)
+    return;
+
+  // Check for dynamic indices in the instruction
+  llvm::SmallVector<std::pair<Value *, uint64_t>, 1> dynIndices;
+  std::vector<Value *> indices;
+  for (unsigned i = 1, operandCount = getElemPtr.getNumOperands(); i < operandCount; ++i) {
+    auto index = getElemPtr.getOperand(i);
+    if (!isa<Constant>(index)) {
+      auto indexedTy = GetElementPtrInst::getIndexedType(pointerTy->getPointerElementType(), indices);
+      if (auto vectorTy = dyn_cast<FixedVectorType>(indexedTy)) {
+        dynIndices.push_back({index, vectorTy->getNumElements()});
+      } else if (auto arrayTy = dyn_cast<ArrayType>(indexedTy)) {
+        dynIndices.push_back({index, arrayTy->getNumElements()});
+      } else {
+        llvm_unreachable("Unsupported indexed type for bounds checking");
+      }
+      break;
+    }
+    indices.push_back(index);
+  }
+  if (dynIndices.empty())
+    return;
+  m_getElementPtrInsts.push_back({&getElemPtr, dynIndices});
+}
+
+// =====================================================================================================================
+// Checks whether each dynamic index in the specified "getelementptr" instruction is lower than its allowed upper bound.
+//
+// @param info : "GetElementPtr" instruction and its dynamic indinces
+// @returns : True of the code was changed or false if unmodified
+bool PatchBoundsCheckMemory::addBoundsCheck(GetElemPtrInfo &info) {
+  auto &getElemPtr = *info.getElemPtr;
+
+  // Copy users
+  std::vector<User *> users;
+  for (auto user : getElemPtr.users()) {
+    // Look through bitcasts
+    Instruction *inst = cast<Instruction>(user);
+    if (isa<BitCastInst>(inst)) {
+      for (auto user : inst->users())
+        users.push_back(user);
+    } else {
+      users.push_back(user);
+    }
+  }
+
+  if (users.empty())
+    return false;
+
+  // Insert after getElemPtr
+  m_builder->SetInsertPoint(getElemPtr.getParent(), ++getElemPtr.getIterator());
+
+  // Get condition for out-of-bounds
+  Value *inBounds = nullptr;
+  for (auto p : info.dynIndices) {
+    auto indexTy = cast<IntegerType>(p.first->getType());
+    auto curInBounds =
+        m_builder->CreateCmp(CmpInst::ICMP_ULT, p.first, m_builder->getIntN(indexTy->getScalarSizeInBits(), p.second));
+    if (inBounds)
+      inBounds = m_builder->CreateAnd(inBounds, curInBounds);
+    else
+      inBounds = curInBounds;
+  }
+
+  // Only execute loads and stores if in-bounds
+  for (auto user : users) {
+    Instruction *inst = cast<Instruction>(user);
+    auto instBB = inst->getParent();
+    auto inBoundsBB = instBB->splitBasicBlock(inst);
+    auto nextInst = ++inst->getIterator();
+    BasicBlock *continueBB = inBoundsBB->splitBasicBlock(nextInst);
```

Comment:
I think there should be an llvm function that splits a BB and adds a branch for the given condition, no?

Reply:
Oh there is, thanks for the hint.